model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/changes/remote/remover/query/QueryBasedConditionalRemover.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright 2024 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.keycloak.models.sessions.infinispan.changes.remote.remover.query;
+
+import java.lang.invoke.MethodHandles;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+
+import org.infinispan.client.hotrod.RemoteCache;
+import org.infinispan.commons.util.concurrent.AggregateCompletionStage;
+import org.jboss.logging.Logger;
+import org.keycloak.models.sessions.infinispan.changes.remote.remover.ConditionalRemover;
+
+/**
+ * An implementation of {@link ConditionalRemover} that uses the delete statement to remove entries from a
+ * {@link RemoteCache}.
+ * <p>
+ * This class is generic and requires the concrete implementation to provide the entity, the condition clause and the
+ * parameters.
+ *
+ * @param <K> The key's type stored in the {@link RemoteCache}.
+ * @param <V> The value's type stored in the {@link RemoteCache}.
+ */
+abstract class QueryBasedConditionalRemover<K, V> implements ConditionalRemover<K, V> {
+
+    private final static Logger logger = Logger.getLogger(MethodHandles.lookup().lookupClass());
+
+    private static final String QUERY_FMT = "DELETE FROM %s WHERE %s";
+
+    @Override
+    public void executeRemovals(RemoteCache<K, V> cache, AggregateCompletionStage<Void> stage) {
+        if (isEmpty()) {
+            return;
+        }
+        // TODO add async method to Infinispan API
+        stage.dependsOn(CompletableFuture.runAsync(() -> executeDeleteStatement(cache)));
```

Comment:
We can use the Infinispan blocking manager. 
It can cause issues with virtual threads. I discovered the hard way that `CompletableFuture.get()` pins the virtual thread in the carrier thread; if all carrier threads are blocked, the system stalls. 
Said that, I don't expect realms to be removed very often and it is probably a good idea to do the change.

Reply:
Added an executor.